moquette-0.16/broker/src/main/java/io/moquette/broker/security/AuthorizationsCollector.java

```diff
@@ -0,0 +1,161 @@
+    private Map<String, List<Authorization>> m_userAuthorizations = new HashMap<>();
+    private boolean m_parsingUsersSpecificSection;
+    private boolean m_parsingPatternSpecificSection;
+    private String m_currentUser = "";
+
+    static final AuthorizationsCollector emptyImmutableCollector() {
+        AuthorizationsCollector coll = new AuthorizationsCollector();
+        coll.m_globalAuthorizations = Collections.emptyList();
+        coll.m_patternAuthorizations = Collections.emptyList();
+        coll.m_userAuthorizations = Collections.emptyMap();
+        return coll;
+    }
+
+    void parse(String line) throws ParseException {
+        Authorization acl = parseAuthLine(line);
+        if (acl == null) {
+            // skip it's a user
+            return;
+        }
+        if (m_parsingUsersSpecificSection) {
+            // TODO in java 8 switch to m_userAuthorizations.putIfAbsent(m_currentUser, new
+            // ArrayList());
+            if (!m_userAuthorizations.containsKey(m_currentUser)) {
+                m_userAuthorizations.put(m_currentUser, new ArrayList<Authorization>());
+            }
+            List<Authorization> userAuths = m_userAuthorizations.get(m_currentUser);
+            userAuths.add(acl);
+        } else if (m_parsingPatternSpecificSection) {
+            m_patternAuthorizations.add(acl);
+        } else {
+            m_globalAuthorizations.add(acl);
+        }
+    }
+
+    protected Authorization parseAuthLine(String line) throws ParseException {
+        String[] tokens = line.split("\\s+");
+        String keyword = tokens[0].toLowerCase();
+        switch (keyword) {
+            case "topic":
+                return createAuthorization(line, tokens);
+            case "user":
+                m_parsingUsersSpecificSection = true;
+                m_currentUser = tokens[1];
+                m_parsingPatternSpecificSection = false;
+                return null;
+            case "pattern":
+                m_parsingUsersSpecificSection = false;
+                m_currentUser = "";
+                m_parsingPatternSpecificSection = true;
+                return createAuthorization(line, tokens);
+            default:
+                throw new ParseException(String.format("invalid line definition found %s", line), 1);
+        }
+    }
+
+    private Authorization createAuthorization(String line, String[] tokens) throws ParseException {
+        if (tokens.length > 2) {
+            // if the tokenized lines has 3 token the second must be the permission
+            try {
+                Authorization.Permission permission = Authorization.Permission.valueOf(tokens[1].toUpperCase());
+                // bring topic with all original spacing
+                Topic topic = new Topic(line.substring(line.indexOf(tokens[2])));
+
+                return new Authorization(topic, permission);
+            } catch (IllegalArgumentException iaex) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)